sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,315 @@
+    """
+    k-medoids class.
+
+    Parameters
+    ----------
+    n_clusters : int, optional, default: 8
+        How many medoids. Must be positive.
+
+    distance_metric : string, optional, default: 'euclidean'
+        What distance metric to use.
+
+    clustering : {'pam'}, optional, default: 'pam'
+        What clustering mode to use.
+
+    init : {'random', 'heuristic'}, optional, default: 'heuristic'
+        Specify medoid initialization.
+
+    max_iter : int, optional, default : 300
+        Specify the maximum number of iterations when fitting.
+
+    random_state : int, optional, default: None
+        Specify random state for the random number generator.
+    """
+
+    # Supported clustering methods
+    CLUSTERING_METHODS = ['pam']
+
+    # Supported initialization methods
+    INIT_METHODS = ['random', 'heuristic']
+
+    def __init__(self, n_clusters=8, distance_metric='euclidean',
+                 clustering_method='pam', init='heuristic',
+                 max_iter=300, random_state=None):
+
+        self.n_clusters = n_clusters
+
+        self.distance_metric = distance_metric
+
+        self.init = init
+
+        self.max_iter = max_iter
+
+        self.clustering_method = clustering_method
+
+        self.random_state = random_state
+
+    def _check_init_args(self):
+
+        # Check n_clusters
+        if self.n_clusters is None or self.n_clusters <= 0 or \
+                not isinstance(self.n_clusters, int):
+            raise ValueError("n_clusters has to be nonnegative integer")
+
+        # Check distance_metric
+        if callable(self.distance_metric):
+            self.distance_func = self.distance_metric
+        elif self.distance_metric in PAIRWISE_DISTANCE_FUNCTIONS:
+            self.distance_func = \
+                PAIRWISE_DISTANCE_FUNCTIONS[self.distance_metric]
+        else:
+            raise ValueError("distance_metric needs to be " +
+                             "callable or one of the " +
+                             "following strings: " +
+                             "%s. Instead, '%s' was given"
+                             % (PAIRWISE_DISTANCE_FUNCTIONS.keys(),
+                                self.distance_metric))
+
+        # Check clustering_method
+        if self.clustering_method not in self.CLUSTERING_METHODS:
+            raise ValueError("clustering must be one of the following: %s"
+                             % self.CLUSTERING_METHODS)
+
+        # Check init
+        if self.init not in self.INIT_METHODS:
+            raise ValueError("init needs to be one of " +
+                             "the following: " +
+                             "%s" % self.INIT_METHODS)
+
+        # Check random state
+        self.random_state_ = check_random_state(self.random_state)
+
+    def fit(self, X, y=None):
+        """Fit K-Medoids to the provided data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape=(n_samples, n_features)
+
```

Comment:
Hm, k-means doesn't have a description for X either, I thought it'q quite self descriptive. I've added a simple description.

Would you expect k-means to have one as well? I can add that to this PR if the one I gave is ok and if this PR is the right place for that. Would be good to keep it consistent.